### PR TITLE
WIP: Support for the ScikitLearn API

### PR DESCRIPTION
**Pull request by cstjean**
I followed your suggestion. The result is not the prettiest, but it works. If it looks reasonable, I'll add the remaining functions and support NMF/KMeans/etc. 

I'd like to understand why you copy `Regularizer` and `Loss` objects. Are they modified anywhere?

### CODE REVIEW

**Comment by madeleineudell** [comment]
I copy the `Regularizer` and `Loss` objects in case the user chooses to scale the losses or regularizers for different columns differently. These objects include a `.scale` field that scales the regularizer or loss, so that, say, fitting column 1 poorly can be penalized more than fitting column 2 poorly.

Your solution for passing in an initialization function is quite nice.

Why do you default to using k=-1?

**Comment by cstjean** [comment]
> I copy the Regularizer and Loss objects in case the user chooses to scale the losses or regularizers for different columns differently. These objects include a .scale field that scales the regularizer or loss, so that, say, fitting column 1 poorly can be penalized more than fitting column 2 poorly.

That's a great level of flexibility.

Normally, `ScikitLearnBase.get_params` should return a Symbol->Value dictionary of all hyperparameters, so that we can do a grid search to find the best value. It would be nice to expose the regularizer's `scale`, but that only works if it's uniform for all rows/columns, otherwise it's too hairy for this interface.

> Why do you default to using k=-1?

The default in scikit-learn-python is `k = number of columns in input matrix`, but we don't have `A` in the constructor, so this is a flag value. It's changed in `fit!`. Do you have a better suggestion?

**Comment by madeleineudell** [comment]
There are indeed quite a large number of hyperparameters: scales for all of the losses and regularizers, and the choice of functional forms for the losses and regularizers (eg QuadraticLoss vs HingeLoss) can all be considered hyperparameters, in addition to the rank `k`. I'd suggest that `ScikitLearnBase.get_params` return the rank `k` and a single scaling parameter that controls the scales of all regularizers (and so, the scaling of the regularizer relative to the loss). Grid search over two parameters is not too bad; grid search over many more is intractable. (Or, a research project!)

Defaulting to `k=-1` seems fine, then; that makes it clear the initial value is not meant to be used.

Would you like to implement a `get_params` method or shall I go ahead and merge?

**Comment by cstjean** [comment]
I'd like to take a stab at get_params, as well as write
documentation/docstrings. I'll take off the WIP when it's ready.

> Grid search over two parameters is not too bad; grid search over many
more is intractable.

I've been thinking about how to improve on that given Julia's ecosystem.
Some people use Gaussian Processes for hyperparameter tuning, it would be
neat to support that.

**Comment by madeleineudell** [comment]
Great, I'll hold off on merging.

Autotuning with GPs would be great! Certainly better than grid search. But my guess is it would still require a large number of calls to the fitting routines --- and you'd have to be quite careful to prevent overfitting.
